### PR TITLE
[lib] fixes #1568 - Add swig files for creating libraries in other languages, tested with python.

### LINKED ISSUE #1568: SWIG files for libskycoin API
Add SWIG interface files in a brand new `lib/swig` folder for use during generation of client libraries for programming languages .

### PR DESCRIPTION
**Pull request by stdevAlDen**
…her languages using swig.

Fixes #1568

Changes:
- Add SWIG .i files for compiling extensions for script languages. Tested with python.

Does this change need to mentioned in CHANGELOG.md?
- Yes

### CODE REVIEW

**Comment by olemis** [comment]
@stdevEclipse pls use `[swig]` prefix . `pyskcoin` is just an instance of a client library that will use Swig bindings .

**Comment by stdevEclipse** [comment]
Got it

File: lib/swig/goslices2.i [hunk 1]
```diff
@@ -0,0 +1,101 @@
+
+%inline {
+	void  __copyToSecKeys(cipher_SecKeys* secKeys, cipher__SecKey* data, int count){
+		secKeys->count = count;
+		for(int i = 0; i < count; i++){
```
[the remark this answers is not in the document]
Comment (stdevEclipse) [reply lib/swig/goslices2.i]: OK. I see it. Copy is not necessary at all. I just realized that.

File: lib/swig/goslices2.i [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+		return result;
+	}
+}
+
+%rename(SKY_cipher_PubKeySlice_Len) wrap_SKY_cipher_PubKeySlice_Len;
+%inline {
+	//[]PubKey
+	GoUint32 wrap_SKY_cipher_PubKeySlice_Len(cipher_PubKeys* pubKeys){
+		GoSlice_ data;
+		__copyFromPubKeys(pubKeys, &data);
```
Comment (olemis) [inline lib/swig/goslices2.i]: I am not sure this is needed at all .

File: lib/swig/goslices2.i [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+		free(data.data);
+		return result;
+	}
+}
+
+%rename(SKY_cipher_PubKeySlice_Less) wrap_SKY_cipher_PubKeySlice_Less;
+%inline {
+	GoUint32 wrap_SKY_cipher_PubKeySlice_Less(cipher_PubKeys* pubKeys, GoInt p1, GoInt p2){
+		GoSlice_ data;
+		__copyFromPubKeys(pubKeys, &data);
```
Comment (olemis) [inline lib/swig/goslices2.i]: I am not sure this is needed at all .

File: lib/swig/goslices2.i [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+%inline {
+	GoUint32 wrap_SKY_cipher_GenerateDeterministicKeyPairsSeed(GoSlice seed, GoInt n, coin__UxArray* newSeed, cipher_SecKeys* secKeys){
+		if( n > MAX_ARRAY_LENGTH_WRAP )
+			return -1;
+		GoSlice_ data;
+		data.data = NULL;
+		data.len = 0;
+		data.cap = 0;
+		GoUint32 result = SKY_cipher_GenerateDeterministicKeyPairsSeed(seed, n, newSeed, &data);
+		__copyToSecKeys(secKeys, data.data, data.len);
```
Comment (olemis) [inline lib/swig/goslices2.i]: I am not sure this is needed at all .

File: lib/swig/goslices2.i [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+		free(data.data);
+		return result;
+	}
+}
+
+%rename(SKY_cipher_PubKeySlice_Swap) wrap_SKY_cipher_PubKeySlice_Swap;
+%inline {
+	GoUint32 wrap_SKY_cipher_PubKeySlice_Swap(cipher_PubKeys* pubKeys, GoInt p1, GoInt p2){
+		GoSlice_ data;
+		__copyFromPubKeys(pubKeys, &data);
```
Comment (olemis) [inline lib/swig/goslices2.i]: I am not sure this is needed at all .

File: lib/swig/gostrings.i [hunk 6]
```diff
@@ -0,0 +1,38 @@
+typedef struct{
+	char* p;
+	int   n;
+} GoString;
```
Comment (olemis) [inline lib/swig/gostrings.i]: Do not repeat this . Include existing header file

File: lib/swig/gostrings.i [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+typedef struct{
+	char* p;
+	int   n;
+} GoString_;
```
Comment (olemis) [inline lib/swig/gostrings.i]: Do not repeat this . Include existing header file

File: lib/swig/goslices2.i [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+%inline {
+	GoUint32 wrap_SKY_cipher_GenerateDeterministicKeyPairs(GoSlice seed, GoInt n, cipher_SecKeys* secKeys){
+		if( n > MAX_ARRAY_LENGTH_WRAP )
+			return -1;
+		GoSlice_ data;
+		data.data = NULL;
+		data.len = 0;
+		data.cap = 0;
+		GoUint32 result = SKY_cipher_GenerateDeterministicKeyPairs(seed, n, &data);
+		__copyToSecKeys(secKeys, data.data, data.len);
```
Comment (olemis) [inline lib/swig/goslices2.i]: I am not sure this is needed at all .

File: lib/swig/goslices2.i [hunk 9]
```diff
@@ -0,0 +1,101 @@
+
```
Comment (olemis) [inline lib/swig/goslices2.i]: rename file `goslices2.i` => `skycoin.mem.i`

File: lib/swig/gostrings.i [hunk 10]
```diff
@@ -0,0 +1,38 @@
+typedef struct{
```
Comment (olemis) [inline lib/swig/gostrings.i]: everything in this file include it in `golang.core.i`

File: lib/swig/handletypemaps.i [hunk 11]
```diff
@@ -0,0 +1,30 @@
+/**
```
Comment (olemis) [inline lib/swig/handletypemaps.i]: everything in this file include it in `skycoin.mem.i` .

File: lib/swig/structs.i [hunk 12]
```diff
@@ -0,0 +1,252 @@
+typedef struct{
+} encrypt__Sha256Xor;
+
+typedef struct{
+    GoInt_ N;
+    GoInt_ R;
+    GoInt_ P;
+    GoInt_ KeyLen;
+} encrypt__ScryptChacha20poly1305;
```
Comment (olemis) [inline lib/swig/structs.i]: Do not duplicate these . Use [SWIG `%include` directive](http://www.swig.org/Doc1.3/Preprocessor.html#Preprocessor_nn2) .

Put this in `skycoin.cipher.encrypt.i`

File: lib/swig/structs.i [hunk 13]
```diff
@@ -0,0 +1,252 @@
+typedef struct{
```
Comment (olemis) [inline lib/swig/structs.i]: Do not duplicate these . Use [SWIG `%include` directive](http://www.swig.org/Doc1.3/Preprocessor.html#Preprocessor_nn2) .

Put this in `skycoin.cipher.encrypt.i`

File: lib/swig/structs.i [hunk 14]
```diff
@@ -0,0 +1,252 @@
...
+    GoInt_ N;
+    GoInt_ R;
+    GoInt_ P;
+    GoInt_ KeyLen;
+} encrypt__ScryptChacha20poly1305;
+
+typedef struct{
+    GoUint32_  n[10];
+} secp256k1go__Field;
```
Comment (olemis) [inline lib/swig/structs.i]: Do not duplicate these . Use [SWIG `%include` directive](http://www.swig.org/Doc1.3/Preprocessor.html#Preprocessor_nn2) .

Put this in `skycoin.cipher.secp256k1go.i`

File: lib/swig/structs.i [hunk 15]
```diff
@@ -0,0 +1,252 @@
...
+
+typedef struct{
+	cipher_SecKey data[MAX_ARRAY_LENGTH_WRAP];
+	int count;
+} cipher_SecKeys;
+
+typedef struct{
+	cipher_PubKey data[MAX_ARRAY_LENGTH_WRAP];
+	int count;
+} cipher_PubKeys;
```
Comment (olemis) [inline lib/swig/structs.i]: Do not duplicate these . Use [SWIG `%include` directive](http://www.swig.org/Doc1.3/Preprocessor.html#Preprocessor_nn2) .

Put definitions above in `skycoin.cipher.crypto.i`

File: lib/swig/simpletypes.i [hunk 16]
```diff
@@ -0,0 +1,21 @@
+%apply int* OUTPUT {GoInt*}
```
Comment (olemis) [inline lib/swig/simpletypes.i]: please move this onto `golang.core.i`

File: lib/swig/goslices2.i (hunk 1 above)
[the remark this answers is not in the document]
Comment (olemis) [reply lib/swig/goslices2.i]: 👍

File: lib/swig/extend.i [hunk 17]
```diff
@@ -0,0 +1,19 @@
+%extend cipher__Address {
```
[the remark this answers is not in the document]
Comment (olemis) [reply lib/swig/extend.i]: @stdevEclipse if u can `%include` header files in `include/` then do all this in a single interface file with all rules needed . If you have to duplicate struct declarations then use names similar to the ones of the generated header files locate in `include/` folder .

**Comment by olemis** [comment]
needs review after merging #1639 .

File: lib/swig/includes.i [hunk 18]
```diff
@@ -0,0 +1,7 @@
+//Apply strictly to python
```
Comment (olemis) [inline lib/swig/includes.i]: @stdevHan is this specific to Python ?

Comment (stdevHan) [reply lib/swig/includes.i]: @olemis  there are only things that need language exclusion

File: lib/swig/dynamic/golang.cgo.i [hunk 19]
```diff
@@ -0,0 +1,96 @@
+/*GoSlice in typemap*/
+%typemap(in) GoSlice {
+	char* buffer = 0;
+	size_t size = 0;
+	int res = SWIG_AsCharPtrAndSize( $input, &buffer, &size, 0 );
```
Comment (olemis) [inline lib/swig/dynamic/golang.cgo.i]: This whole file , even if it includes `SWIG_AsCharPtrAndSize`, should be implemented in `lib/swig/common` . Should a language be missing implementation of string typemap fragments then they should be added in client library repository .

**Pull request merged**